projects/storefrontlib/shared/components/captcha/captcha.service.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * SPDX-FileCopyrightText: 2023 SAP Spartacus team <spartacus-team@sap.com>
+ * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Injectable, OnDestroy } from '@angular/core';
+import {
+  BaseSiteService,
+  CaptchaConfig,
+  LanguageService,
+  ScriptLoader,
+  SiteAdapter,
+} from '@spartacus/core';
+import { forkJoin, map, Observable, ReplaySubject, Subscription } from 'rxjs';
+import { concatMap, take } from 'rxjs/operators';
+import { CaptchaProvider, RenderParams } from './captcha.model';
+import { CaptchaApiConfig } from './captcha-api-config';
+
+/**
+ * Global function to be passes as "onload" url param for captcha <script>, to be
+ * triggered once script and dependencies are loaded
+ */
+declare global {
+  interface Window {
+    onCaptchaLoad: () => void;
+  }
+}
+
+@Injectable({
+  providedIn: 'root',
+})
+export abstract class CaptchaService implements CaptchaProvider, OnDestroy {
+  protected token: string;
+  protected subscription = new Subscription();
+  protected captchaConfigSubject$ = new ReplaySubject<CaptchaConfig>(1);
+  protected captchaConfig: CaptchaConfig;
+
+  constructor(
+    protected adapter: SiteAdapter,
+    protected apiConfig: CaptchaApiConfig,
+    protected languageService: LanguageService,
+    protected scriptLoader: ScriptLoader,
+    protected baseSiteService: BaseSiteService
+  ) {
+    this.initialize();
+  }
+
+  /**
+   * Retrieve captcha configuration from the backend, and if enabled
+   * call @function loadScript with active language
+   */
+  initialize(): void {
+    this.subscription.add(
+      this.fetchCaptchaConfigFromServer().subscribe((config) => {
+        if (config?.enabled) {
+          this.captchaConfig = config;
+          this.loadResource();
+        } else {
+          this.captchaConfigSubject$.next({ enabled: false });
+        }
+      })
+    );
+  }
+
+  fetchCaptchaConfigFromServer(): Observable<CaptchaConfig> {
+    return forkJoin([
+      this.languageService.getActive().pipe(take(1)),
+      this.baseSiteService.getActive().pipe(
+        concatMap((value) => this.adapter.loadBaseSite(value)),
+        take(1)
+      ),
+    ]).pipe(map((result) => result[1]?.captchaConfig as CaptchaConfig));
+  }
+
+  getCaptchaConfig(): Observable<CaptchaConfig> {
+    return this.captchaConfigSubject$.asObservable();
+  }
+
+  /**
+   * Trigger rendering function configured in CaptchaApiConfig
+   * @param {HTMLElement} elem - HTML element to render captcha widget within.
+   */
+  abstract renderCaptcha(renderParams: RenderParams): Observable<string>;
```

Comment:
Looks like customers will have to implement this method themselves in a custom sub-class. (There's a default implementation in `MockCaptchaService`, but that does not seem to do anything that would be useful for integrating with a real Captcha service provider.)

Imho this abstract method deserves better documentation. From a security perspective, the docs should clearly mention that implementors are responsible for preventing XSS when modifying the `element` in the `RenderParams`.

Reply:
These help portal docs mention how to implement a custom `CaptchaService`:

https://help.sap.com/docs/INTERNAL_COMPOSABLE_STOREFRONT/eaef8c61b6d9477daf75bff9ac1b7eb4/4605061113284c2c92cc558600140aaa.html?locale=en-US&state=DRAFT#loiofea8e9b575ef447480ee90b5c1b2

But they don't even mention how to implement the `renderCaptcha` method. Not very helpful.

Reply:
It's complex for different Captcha provider may has different requirements, some of them need a click, others may need input. It's little complex to have sth in here. Maybe we can write a blog for it.

Reply:
Sure, the details differ, but the concepts are often similar. And imho the generic captcha plugin-point on the OCC side and this `CaptchaService` are suitable for integration with many different captcha service providers.

But since we're leaving the captcha service integration to the customer, we should provide better documentation. And ideally provide working code samples for at least one or two captcha service providers. (This should include reCAPTCHA, since that's what the classic accelerators supported.)

Even if we don't ship an out-of-the-box captcha service integration with Spartacus, it would be great to have e2e tests with at least one real-life captcha service. Just to demonstrate feasibility of the framework that we provide.